%% 1. Load data and filter out non-P301S datasets
clear; clc; rng(0);
matdir = '/Users/justintorok/Documents/MATLAB/CellTypeVulnerability_Project/CellTypeVulnerability/MatFiles';
load([matdir filesep 'Mouse_Tauopathy_Data.mat'],'mousedata_struct'); % code below already run
datsetnames = fieldnames(mousedata_struct);

%% 2. Univariate analysis, cell-type-based vulnerability
%% 2.1 Correlation box plots, cell types, end timepoint
addpath('/Users/justintorok/Documents/MATLAB/CellTypeVulnerability_Project/CellTypeVulnerability/Code/violin');
datapath = '/Users/justintorok/Documents/MATLAB/CellTypeVulnerability_Project/Large_MatFiles';
figdirectory = '/Users/justintorok/Documents/MATLAB/CellTypeVulnerability_Project/Figures/DraftFigs';

ctdataset = 'Yao';
if strcmp(ctdataset,'Tasic')
    load([datapath filesep 'Tasic_Dependencies.mat']);
    subclasses = classkey;
    strepper1 = @(x) strrep(x,'_','-');
    subclasses = cellfun(strepper1,subclasses,'UniformOutput',0);
    strepper2 = @(x) strrep(x,'Macro','Micro');
    subclasses = cellfun(strepper2,subclasses,'UniformOutput',0);
    strepper3 = @(x) strrep(x,'L2-3','L2/3');
    subclasses = cellfun(strepper3,subclasses,'UniformOutput',0);
elseif strcmp(ctdataset,'Yao')
    load([datapath filesep 'Yao_Dependencies.mat']);
    subclasses = classkey;
end

plotting = 1;
savenclose = 0;
corrtype = 'Pearson';
ctmat = outstruct.Bmeans;
t = 'end';
cmap = [[0 0.75 1]; [1 0 0.5]];
figtype = 'Yao';
[corrmat_t,logpmat_t] = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,ctmat,corrtype);
if plotting
    CorrelationBarPlot(corrmat_t,datsetnames,subclasses,corrtype,...
        cmap,figtype,savenclose,figdirectory);
    PValuePlot(logpmat_t,datsetnames,subclasses,corrtype,...
        copper(1000),cmap,savenclose,figdirectory);
    [classes,classmeans,pvals] = ClassViolinPlot(corrmat_t,subclasses,corrtype,...
        cmap,savenclose,figdirectory);
end

%% 2.2 Top 2 Cell Types, Brainframe 
% Requires the Brainframe repo: https://github.com/Raj-Lab-UCSF/Brainframe
brainframedir = '/Users/justintorok/Documents/MATLAB/Brainframe-Dev/Brainframe';
addpath(brainframedir)
corrmat_end_mean = mean(corrmat_t);
[~,maxind] = max(corrmat_end_mean);
[~,minind] = min(corrmat_end_mean);
toptypeinds = [maxind,minind];

plotting = 1;
savenclose = 0;
if length(subclasses) == 42 % Yao
    if plotting
        % Define colormap based on cell class
        nonneuronal_inds = ismember(subclasses,{'Endo','Astro','Oligo',...
            'Micro-PVM','SMC-Peri','VLMC'});
        gaba_inds = ismember(subclasses,{'Lamp5','Sncg','Meis2','CR',...
            'Pvalb','Sst','Sst Chodl','Vip'});
        ctxtest = @(x) strcmp(x(end),'X');
        glutctx_inds = logical(cell2mat(cellfun(ctxtest,subclasses,'UniformOutput',false)));
        gluthipp_inds = ~logical(nonneuronal_inds + gaba_inds + glutctx_inds);
        indcell = {glutctx_inds,gluthipp_inds,gaba_inds,nonneuronal_inds};
        indtest = glutctx_inds + 2*gluthipp_inds + 3*gaba_inds + 4*nonneuronal_inds;
        cmap_col = [[0 0.75 1]; [1 0 0.5]];
        cmap = twocolor(cmap_col(1,:),cmap_col(2,:),length(indcell));
        voxthreshes = [0.6,0.6]; % Brainframe parameter
        
        for i = 1:length(toptypeinds)
            ctdata = outstruct.corrB(:,toptypeinds(i));
            newVoxMap = zeros(size(GENGDmod));
            newVoxMap(nonzerovox) = ctdata;
            datinput = imresize3(newVoxMap,[133 81 115]);
            datinput(datinput < 0) = 0;
            col_base = cmap(indtest(toptypeinds(i)),:);
            col_min = (col_base + 1)/2;
            nbin = 10;
            cmap_i = twocolor(col_min,col_base,nbin);
            input_struct = brainframe_inputs_mouse(brainframedir,'data',datinput,...
                                                        'voxthresh',voxthreshes(i),...
                                                        'nbin',nbin,...
                                                        'voxUreg',0,...
                                                        'xfac',0.02,...
                                                        'pointsize',0.1,...
                                                        'bgcolor','w',...
                                                        'img_format','tiffn',...
                                                        'cmap',cmap_i,...
                                                        'regsUbins',0,...
                                                        'img_directory',figdirectory,...
                                                        'img_labels',['Yao_' subclasses{toptypeinds(i)}],...
                                                        'img_renderer','painters',...
                                                        'savenclose',savenclose);
            brainframe(input_struct);
        end
    else
        % % Finish later for Tasic/Zeisel   
        %     nonneuronal_inds = ismember(subclasses,{'Endo','Astro','Oligo',...
        %         'Macro'});
        %     gaba_inds = ismember(subclasses,{'Lamp5','Sncg','Meis2','CR',...
        %         'Pvalb','Sst','Serpinf1','Vip'});
        %     ctxtest = @(x) strcmp(x(end),'X');
        %     glutctx_inds = logical(cell2mat(cellfun(ctxtest,subclasses,'UniformOutput',false)));
        %     gluthipp_inds = ~logical(nonneuronal_inds + gaba_inds + glutctx_inds);
        %     indcell = {glutctx_inds,gluthipp_inds,gaba_inds,nonneuronal_inds};
        %     indtest = glutctx_inds + 2*gluthipp_inds + 3*gaba_inds + 4*nonneuronal_inds;
    end
end

%% 3. Gene-based vulnerability
%% 3.1 Gene expression distance from CA1-ProS
plotting = 1;
savenclose = 0;
genemat_allgenes = genevct_allgenes;
genemat_MRx3 = genevct(sort(geneinds(1:outstruct.nGen)),:);
ctmat = outstruct.Bmeans;
corrtype = 'Pearson';
genemattype = 'MRx3';
neuronly = [0,1];
usewls = [0,1];
t = 'end';
corrmat_t = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,ctmat,corrtype);
cmap = [[0 0.75 1]; [1 0 0.5]];

if plotting
    for i = 1:length(neuronly)
        for j = 1:length(usewls)
            GeneExpressionDistance_vs_Tau_Plot(genemat_allgenes,genemat_MRx3,...
                genemattype,subclasses,neuronly(i),corrmat_t,corrtype,cmap,usewls(j),...
                savenclose,figdirectory);
        end
    end
end

%% 3.2 Genetic vulnerability signature, end timepoint, overall across cell types
thresh = 0.1;
n_g = 50;
t = 'end';
stdweight = 1;
corrtype = 'Pearson';
corrmat_t = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,ctmat,corrtype);
[vuln_genes,vscore] = VulnerabilityGeneSignatureIdentifier(corrmat_t,genevct,gene_names,...
    n_g,thresh,stdweight);
genelist_inds = ismember(gene_names,vuln_genes);
genemat_list = voxvgene(:,genelist_inds);
[~,genemat_list] = Voxel_To_Region(genemat_list,matdir);
corrmat_t_gene = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,genemat_list,corrtype);

plotting = 1;
savenclose = 1;
cmap = [[1 0 0]; [1 0.75 0]];
if stdweight == 0
    figtype = ['CTVuln_Genes_' num2str(n_g) '_Mean'];
elseif stdweight == 1
    figtype = ['CTVuln_Genes_' num2str(n_g) '_CoV'];
else
    figtype = ['CTVuln_Genes_' num2str(n_g) '_StdWeight' num2str(stdweight)];
end

if plotting
    CorrelationBarPlot(corrmat_t_gene,datsetnames,vuln_genes,corrtype,...
        cmap,figtype,savenclose,figdirectory);
end

%% 3.3 Vulnerability from common AD genes
corrtype = 'Pearson';
load([matdir filesep 'AD_Genes.mat'],'genelist');
t = 'end';
genelist_inds = zeros(1,length(genelist));
for i = 1:length(genelist_inds)
    genelist_inds(i) = find(ismember(gene_names,genelist{i}));
end
genemat_list = voxvgene(:,genelist_inds);
[~,genemat_list] = Voxel_To_Region(genemat_list,matdir);
corrmat_t_gene = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,genemat_list,corrtype);

plotting = 1;
savenclose = 1;
cmap = [[1 0 0]; [1 0.75 0]];
figtype = 'AD_Genes';
if plotting
    CorrelationBarPlot(corrmat_t_gene,datsetnames,genelist,corrtype,...
        cmap,figtype,savenclose,figdirectory);
end

%% 3.4 Genetic vulnerability vs. differential expression
t = 'end';
corrmat_t_ct = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,ctmat,corrtype);
[~,genemat_list] = Voxel_To_Region(voxvgene,matdir);
corrmat_t_gene = CorrelationCalculator_singletimepoint(datsetnames,t,...
    mousedata_struct,genemat_list,corrtype);
thresh = 0;
n_g = 100;
stdweight = 0;

plotting = 1;
savenclose = 0;
if plotting
    GeneVuln_vs_DE_Plot(corrmat_t_ct,corrmat_t_gene,genevct,gene_names,...
        n_g,thresh,stdweight);
end
% corrmat_t_gene_mean = mean(corrmat_t_gene);
% [~,sortinds] = sort(corrmat_t_gene_mean,'descend');
% vuln_genes = gene_names(sortinds(1:100));
% go_genes = [go_genes, {'Gene Vuln'; vuln_genes}];

%% 4. Internal structure of data
%% 4.1 Cross-correlation matrices with hierarchical clustering
plotting = 1;
savenclose = 0;
corrtype = 'Spatial';
clusttype = 'Spatial';
linktype = 'ward';
disttype = 'euclidean';
uselab = 0;
genemat_allgenes = genevct_allgenes;
genemat_MRx3 = genevct(sort(geneinds(1:outstruct.nGen)),:);
ctmat = outstruct.Bmeans;
cmap_label = [[0 0.75 1]; [1 0 0.5]];
cmap = [];

if plotting
    ClusteringHeatmap(genemat_allgenes,genemat_MRx3,ctmat,subclasses,...
        ctdataset,corrtype,clusttype,cmap,cmap_label,disttype,linktype,...
        uselab,savenclose,figdirectory);
end

%% 4.2 Cross-correlation matrices of data
plotting = 1;
savenclose = 0;
cmap = [[ones(500,1), 0.9*linspace(1,0.5,500).', 0.9*linspace(1,0,500).'];...
                    [ones(500,1), linspace(0.5,0,500).', 0*ones(500,1)]].^(1/2);

if plotting
    TauPathologyHeatmap(datsetnames,mousedata_struct,cmap,savenclose,figdirectory);
end

%% 5. Multivariate analyses
%% 5.1 Linear models and frequency bar plots, cell types
plotting = 1;
bicplotting = 0;
savenclose = 0;
usebic = 1;
numsigtypes = 5;
ctdataset = 'Yao';
corrtype = 'Pearson';
ctmat = outstruct.Bmeans;
ts = 1:3;

for i = ts
    t = ts(i);
    [mdls_bic,subclasses_bic] = LinearModelSelection_BIC_alldatasets(datsetnames,t,mousedata_struct,...
        ctmat,subclasses,corrtype,usebic,ctdataset,bicplotting,plotting,[],...
        savenclose,figdirectory);
    [mdls_topN,subclasses_topN] = LinearModelSelection_TopN_alldatasets(datsetnames,t,...
        mousedata_struct,ctmat,subclasses,numsigtypes,corrtype,ctdataset,plotting,...
        [],savenclose,figdirectory);
    [mdls_all,subclasses_all] = LinearModelSelection_TopN_alldatasets(datsetnames,t,...
        mousedata_struct,ctmat,subclasses,length(subclasses),corrtype,ctdataset,plotting,...
        [],savenclose,figdirectory);
    CellTypeFrequencyPlot(subclasses,subclasses_bic,t,ctdataset,1,[[0 0.75 1]; [1 0 0.5]],...
        savenclose,figdirectory);
    CellTypeFrequencyPlot(subclasses,subclasses_topN,t,ctdataset,0,[[0 0.75 1]; [1 0 0.5]],...
        savenclose,figdirectory);
end

%% 5.2 Linear models, AD genes
plotting = 1;
savenclose = 0;
ctdataset = 'Gene';
corrtype = 'Pearson';
load([matdir filesep 'AD_Genes.mat'],'genelist');
ts = 1:3;

genelist_inds = zeros(1,length(genelist));
for i = 1:length(genelist_inds)
    genelist_inds(i) = find(ismember(gene_names,genelist{i}));
end
genemat_list = voxvgene(:,genelist_inds);
[~,genemat_list] = Voxel_To_Region(genemat_list,matdir);

for i = ts
    t = ts(i);
    [mdls_all_gene,subclasses_all_gene] = LinearModelSelection_TopN_alldatasets(datsetnames,t,...
        mousedata_struct,genemat_list,genelist,length(genelist),corrtype,ctdataset,plotting,...
        [],savenclose,figdirectory);
end


%% 6.1 Hurtado per-timepoint
plotting = 1;
bicplotting = 0;
savenclose = 0;
usebic = 1;
ctdataset = 'Yao';
corrtype = 'Pearson';
datset = 'Hurtado';
ctmat = outstruct.Bmeans;
cmap = twocolor([0,1,0.25],[0.5,0,1],length(mousedata_struct.(datset).time_stamps));

[mdls_bic_hurtado,subclasses_bic_hurtado,xcti,Y] = LinearModelSelection_BIC_alltpts(...
    datset,mousedata_struct, ctmat,subclasses,corrtype,usebic,ctdataset,...
    bicplotting,plotting,cmap,savenclose,figdirectory);

%% 6.2 Hurtado pathology glass-brains, space-filling
brainframedir = '/Users/justintorok/Documents/MATLAB/Brainframe-Dev/Brainframe';
addpath(brainframedir)
datset = 'Hurtado';
tpts = mousedata_struct.(datset).time_stamps;
savenclose = 0;
reggroups = zeros(213,1); %Chunk of code to define region_groups
amy = 1:11; cer = 12:23; sub = 24:26; hip = 27:37; hyp = 38:57; 
ncx = 58:95; med = 96:120; mid = 121:141; olf = 142:149; pal = 150:157;
pon = 158:170; str = 171:178; tha = 179:213;
reggroups(amy) = 1; reggroups(cer) = 2; reggroups(sub) = 3; 
reggroups(hip) = 4; reggroups(hyp) = 5; reggroups(ncx) = 6;
reggroups(med) = 7; reggroups(mid) = 8; reggroups(olf) = 9;
reggroups(pal) = 10; reggroups(pon) = 11; reggroups(str) = 12;
reggroups(tha) = 13;
reggroups = [reggroups;reggroups];
cmap = hsv(length(unique(reggroups))); %Creating colormap

for i = 1:length(tpts)
    datinput = mousedata_struct.(datset).data(:,i);
    nany = isnan(datinput);
    datinput(nany) = 0;
    xfac = 15*(max(datinput)/max(mousedata_struct.(datset).data(~nany,1)));
    imglab = sprintf('PathologyBrainframe_Data_%s_t%d_%s',datset,tpts(i),ctdataset);
    input_struct_data = brainframe_inputs_mouse(brainframedir,'data',datinput,...
                                                'voxUreg',1,...
                                                'xfac',xfac,...
                                                'pointsize',15,...
                                                'norm_method','max',...
                                                'bgcolor','w',...
                                                'img_format','tiffn',...
                                                'cmap',cmap,...
                                                'region_groups',reggroups,...
                                                'centered',[0 1],...
                                                'regsUbins',0,...
                                                'img_directory',figdirectory,...
                                                'img_labels',imglab,...
                                                'img_renderer','painters',...
                                                'savenclose',savenclose);
    brainframe(input_struct_data);

    mdlpred = mdls_bic_hurtado{i}.Fitted;
    mdlpred_1 = mdls_bic_hurtado{1}.Fitted;
    mdlpred_norm = (mdlpred - min(mdlpred));
    datinput = zeros(426,1);
    datinput(~nany) = mdlpred_norm;
    xfac = 15*max(mdlpred)/max(mdlpred_1);
    imglab = sprintf('PathologyBrainframe_Predicted_%s_t%d_%s',datset,tpts(i),ctdataset);
    input_struct_pred = brainframe_inputs_mouse(brainframedir,'data',datinput,...
                                                'voxUreg',1,...
                                                'xfac',xfac,...
                                                'pointsize',15,...
                                                'norm_method','max',...
                                                'bgcolor','w',...
                                                'img_format','tiffn',...
                                                'cmap',cmap,...
                                                'region_groups',reggroups,...
                                                'centered',[0 1],...
                                                'regsUbins',0,...
                                                'img_directory',figdirectory,...
                                                'img_labels',imglab,...
                                                'img_renderer','painters',...
                                                'savenclose',savenclose);
    brainframe(input_struct_pred);
end

%% 6.3 Top Hurtado cell types, glass brains
brainframedir = '/Users/justintorok/Documents/MATLAB/Brainframe-Dev/Brainframe';
addpath(brainframedir)
toptypes = subclasses;
for i = 1:length(subclasses_bic_hurtado)
    toptypes = intersect(toptypes,subclasses_bic_hurtado{i});
end
toptypeinds = find(ismember(subclasses,toptypes));

plotting = 1;
savenclose = 0;
if length(subclasses) == 42
    if plotting
        nonneuronal_inds = ismember(subclasses,{'Endo','Astro','Oligo',...
            'Micro-PVM','SMC-Peri','VLMC'});
        gaba_inds = ismember(subclasses,{'Lamp5','Sncg','Meis2','CR',...
            'Pvalb','Sst','Sst Chodl','Vip'});
        ctxtest = @(x) strcmp(x(end),'X');
        glutctx_inds = logical(cell2mat(cellfun(ctxtest,subclasses,'UniformOutput',false)));
        gluthipp_inds = ~logical(nonneuronal_inds + gaba_inds + glutctx_inds);
        indcell = {glutctx_inds,gluthipp_inds,gaba_inds,nonneuronal_inds};
        indtest = glutctx_inds + 2*gluthipp_inds + 3*gaba_inds + 4*nonneuronal_inds;
        cmap_col = [[0 0.75 1]; [1 0 0.5]];
        cmap = twocolor(cmap_col(1,:),cmap_col(2,:),length(indcell));
        voxthreshes = 0.6*ones(1,length(toptypes));
        
        datamat = outstruct.corrB; 
        for i = 1:length(toptypeinds)
            ctdata = outstruct.corrB(:,toptypeinds(i));
            newVoxMap = zeros(size(GENGDmod));
            newVoxMap(nonzerovox) = ctdata;
            datinput = imresize3(newVoxMap,[133 81 115]);
            datinput(datinput < 0) = 0;
            col_base = cmap(indtest(toptypeinds(i)),:);
            col_min = (col_base + 1)/2;
            nbin = 10;
            cmap_i = twocolor(col_min,col_base,nbin);
            labstr = ['Yao_' strrep(subclasses{toptypeinds(i)},' ','_')];
            labstr = strrep(labstr,'/','-');
            input_struct = brainframe_inputs_mouse(brainframedir,'data',datinput,...
                                                        'voxthresh',voxthreshes(i),...
                                                        'nbin',nbin,...
                                                        'voxUreg',0,...
                                                        'xfac',0.02,...
                                                        'pointsize',0.1,...
                                                        'bgcolor','w',...
                                                        'img_format','tiffn',...
                                                        'cmap',cmap_i,...
                                                        'regsUbins',0,...
                                                        'img_directory',figdirectory,...
                                                        'img_labels',labstr,...
                                                        'img_renderer','painters',...
                                                        'savenclose',savenclose);
            brainframe(input_struct);
        end
    end
else
    % % Finish later for Tasic/Zeisel     
    %     nonneuronal_inds = ismember(subclasses,{'Endo','Astro','Oligo',...
    %         'Macro'});
    %     gaba_inds = ismember(subclasses,{'Lamp5','Sncg','Meis2','CR',...
    %         'Pvalb','Sst','Serpinf1','Vip'});
    %     ctxtest = @(x) strcmp(x(end),'X');
    %     glutctx_inds = logical(cell2mat(cellfun(ctxtest,subclasses,'UniformOutput',false)));
    %     gluthipp_inds = ~logical(nonneuronal_inds + gaba_inds + glutctx_inds);
    %     indcell = {glutctx_inds,gluthipp_inds,gaba_inds,nonneuronal_inds};
    %     indtest = glutctx_inds + 2*gluthipp_inds + 3*gaba_inds + 4*nonneuronal_inds;
end

%% 7. Miscellaneous figures
%% 7.1 Plots for schematic
% tSNE
savenclose = 0;
usetSNE = 1;
genelist_schem = sort({'Sst','Pvalb','Rorb','Aqp4','Mapt','App','Apoe'});
ClusteringSchematic(usetSNE,savenclose,figdirectory);
GeneExpressionSchematic(genelist_schem,34,savenclose,datapath,figdirectory);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% %% 4. Ridge-based cell-type subset selection w.r.t. t(end) tau pathology 
% rng(0);
% lambdas = 2.^linspace(0,12,200);
% lambdas_min = zeros(1,length(datsetnames));
% lambdas_1se = lambdas_min;
% sigtypeinds_min = cell(1,length(datsetnames));
% sigtypeinds_1se = sigtypeinds_min; sigtypeinds_0 = sigtypeinds_min;
% sigtypeinds_min_med = cell(1,length(datsetnames));
% sigtypeinds_1se_med = sigtypeinds_min; sigtypeinds_0_med = sigtypeinds_min;
% B_mins = sigtypeinds_min; B_1ses = sigtypeinds_min; B_0s = sigtypeinds_min;
% kfold = 10;
% 
% plotting = 0;
% savenclose = 0;
% if plotting
%     f1 = figure('Units','inches','Position',[0,0,20,10]);
% end
% for i = 1:length(datsetnames)
%     % K-fold cross-validation for ridge regression
%     datsetname = datsetnames{i};
%     fprintf('Finding Optimal Lambda, %s \n',datsetname)
%     Y = mousedata_struct.(datsetname).data;
%     naninds = isnan(Y(:,1));
%     Y_end = Y(~naninds,end);
%     Xct = Yao_nG1300(~naninds,:);
% %     Xct = Tasic_ng606(~naninds,:);
%     Xct_sums = repmat(mean(Xct),size(Xct,1),1);
%     Xct_norm = Xct./Xct_sums;
%     c = cvpartition(length(Y_end),'KFold',kfold);
%     residuals_sq = zeros(kfold,length(lambdas));
%     for k = 1:kfold
%         testinds = test(c,k);
%         Y_k_test = Y_end(testinds); Xct_k_test = Xct_norm(testinds,:);
%         traininds = ~testinds;
%         Y_k_train = Y_end(traininds); Xct_k_train = Xct_norm(traininds,:);
%         B = ridge(Y_k_train,Xct_k_train,lambdas,1);
%         Y_k_test_mat = repmat(Y_k_test,1,length(lambdas));
%         Y_k_test_mat_pred = Xct_k_test * B;
%         Y_k_train_mat_pred = Xct_k_train * B;
%         Y_diff_sq = (Y_k_test_mat - Y_k_test_mat_pred).^2;
%         residuals_sq(k,:) = sum(Y_diff_sq);
%     end
%     % Determination of minimum and minimum + 1se lambda values
%     mean_residuals_sq = mean(residuals_sq);
%     std_residuals_sq = std(residuals_sq);
%     [minmse,minind] = min(mean_residuals_sq);
%     lambdas_min(i) = lambdas(minind);
%     minmse_plus1se = minmse + std_residuals_sq(minind);
%     mean_subtract_abs = abs(mean_residuals_sq - minmse_plus1se);
%     mean_subtract_abs(lambdas < lambdas_min(i)) = Inf;
%     [~,min1seind] = min(mean_subtract_abs);
%     lambdas_1se(i) = lambdas(min1seind);
%     
%     % Bootstrapping for lambda = 0 (OLS), lambda_min, lambda_1se
%     fprintf('Bootstrap-based Coefficient Estimation, %s \n',datsetname)
%     niters = 1000;
%     B_matrix_1se = zeros(size(Xct_norm,2),niters);
%     B_matrix_min = B_matrix_1se;
%     B_matrix_0 = B_matrix_1se;
%     for j = 1:niters
%         traininds = datasample(1:length(Y_end),length(Y_end),'Replace',true);
%         Y_k_train = Y_end(traininds); Xct_k_train = Xct_norm(traininds,:);
%         B_matrix_min(:,j) = ridge(Y_k_train,Xct_k_train,lambdas_min(i),1);
%         B_matrix_1se(:,j) = ridge(Y_k_train,Xct_k_train,lambdas_1se(i),1);
%         B_matrix_0(:,j) = ridge(Y_k_train,Xct_k_train,0,1);
%     end
% 
%     sigthresh = 1; % Defines CI around mean to be +/- sigthresh * std
%     B_mean_min = mean(B_matrix_min,2); B_std_min = std(B_matrix_min,[],2);
%     B_upper_min = B_mean_min + sigthresh*B_std_min; B_lower_min = B_mean_min - sigthresh*B_std_min;
%     sigtypeinds_min{i} = find(sign(B_upper_min) == sign(B_lower_min));
%     B_mins{i} = B_matrix_min;
% 
%     B_mean_1se = mean(B_matrix_1se,2); B_std_1se = std(B_matrix_1se,[],2);
%     B_upper_1se = B_mean_1se + sigthresh*B_std_1se; B_lower_1se = B_mean_1se - sigthresh*B_std_1se;
%     sigtypeinds_1se{i} = find(sign(B_upper_1se) == sign(B_lower_1se));
%     B_1ses{i} = B_matrix_1se;
%     
%     B_mean_0 = mean(B_matrix_0,2); B_std_0 = std(B_matrix_0,[],2);
%     B_upper_0 = B_mean_0 + sigthresh*B_std_0; B_lower_0 = B_mean_0 - sigthresh*B_std_0;
%     sigtypeinds_0{i} = find(sign(B_upper_0) == sign(B_lower_0));
%     B_0s{i} = B_matrix_0;
%     B_mean_0s(i,:) = B_mean_0.';
%     B_means_mat = [B_mean_0,B_mean_min,B_mean_1se];
% 
%     B_med_min = median(B_matrix_min,2); B_std_min = std(B_matrix_min,[],2);
%     B_upper_min = B_med_min + sigthresh*B_std_min; B_lower_min = B_med_min - sigthresh*B_std_min;
%     sigtypeinds_min_med{i} = find(sign(B_upper_min) == sign(B_lower_min));
% %     B_mins{i} = B_matrix_min;
% 
%     B_med_1se = median(B_matrix_1se,2); B_std_1se = std(B_matrix_1se,[],2);
%     B_upper_1se = B_med_1se + sigthresh*B_std_1se; B_lower_1se = B_med_1se - sigthresh*B_std_1se;
%     sigtypeinds_1se_med{i} = find(sign(B_upper_1se) == sign(B_lower_1se));
% %     B_1ses{i} = B_matrix_1se;
%     
%     B_med_0 = median(B_matrix_0,2); B_std_0 = std(B_matrix_0,[],2);
%     B_upper_0 = B_med_0 + sigthresh*B_std_0; B_lower_0 = B_med_0 - sigthresh*B_std_0;
%     sigtypeinds_0_med{i} = find(sign(B_upper_0) == sign(B_lower_0));
% %     B_0s{i} = B_matrix_0;
%     B_meds_mat = [B_med_0,B_med_min,B_med_1se];
% 
% 
%     if plotting
%         figure(f1)
%         subplot(3,4,i); hold on;
%         inds = 1:2:(length(lambdas)-1);
%         errorbar(-log2(lambdas(inds)),mean_residuals_sq(inds),std_residuals_sq(inds),'o'); 
%         plot([-log2(lambdas_min(i)), -log2(lambdas_min(i))],...
%             [min(mean_residuals_sq)-1.1*max(std_residuals_sq),max(mean_residuals_sq)+1.1*max(std_residuals_sq)],...
%             'LineStyle',':','LineWidth',2,'Color','b');
%         plot([-log2(lambdas_1se(i)), -log2(lambdas_1se(i))],...
%             [min(mean_residuals_sq)-1.1*max(std_residuals_sq),max(mean_residuals_sq)+1.1*max(std_residuals_sq)],...
%             'LineStyle',':','LineWidth',2,'Color','g');
%         if i == 1
%             legend({sprintf('%d-fold CV',kfold),'Min MSE', 'Min MSE + 1SE'},'Location','southeast',...
%                 'FontSize',16);
%         end
%         ylim([min(mean_residuals_sq)-1.1*max(std_residuals_sq),max(mean_residuals_sq)+1.1*max(std_residuals_sq)])
%         xlim([-log2(lambdas(end)),-log2(lambdas(1))])
%         xlabel('-log2(\lambda)'); 
%         if i == 1 || i == 5
%             ylabel(sprintf('MSE (%d-fold CV)',kfold)); 
%         end
%         title(datsetlabels{i});
%         set(gca,'FontSize',20,'FontName','Times')
%     end
%     cmap = hsv(length(subclasses));
%     if plotting 
%         figure('Position',[0 0 1000 1500]); 
%         subplot(3,1,1); hold on;
%         b = boxplot(B_matrix_0.',g,'Colors',cmap,'Symbol','');
%         set(b,{'linew'},{1})
%         plot([0,length(classkey)+1],[0,0],'k--','LineWidth',2);
%         set(gca,'XTick',1:length(subclasses));
%         set(gca,'XTickLabel',{})
%         ylim_ = max(abs([max(B_matrix_0(:)),min(B_matrix_0(:))]));
%         set(gca,'YTick',[-ylim_,ylim_])
%         ytickformat('%0.2f')
%         text(-0.45,0,'0','FontSize',16,'FontName','Times')
%         ylim([-1.1*ylim_,1.1*ylim_]);
%         ylabel('Regression Coefficients');
%         xlabel('');
%         set(gca,'TickLength',[0 0])
%         title(['OLS Coefficients, ' datsetlabels{i}],'FontSize',28);
%         set(gca, 'FontSize', 16, 'FontName', 'Times');
%         subplot(3,1,2); hold on;
%         b = boxplot(B_matrix_min.',g,'Colors',cmap,'Symbol','');
%         set(b,{'linew'},{1})
%         plot([0,length(classkey)+1],[0,0],'k--','LineWidth',2);
%         set(gca,'XTick',1:length(subclasses));
%         set(gca,'XTickLabel',{});
%         ylim_ = max(abs([max(B_matrix_min(:)),min(B_matrix_min(:))]));
%         set(gca,'YTick',[-ylim_,ylim_])
%         ytickformat('%0.2f')
%         text(-0.45,0,'0','FontSize',16,'FontName','Times')
%         ylim([-1.1*ylim_,1.1*ylim_]);
%         ylabel('Regression Coefficients');
%         xlabel('');
%         set(gca,'TickLength',[0 0])
%         title(['\lambda_m_i_n Ridge Coefficients, ' datsetlabels{i}],'FontSize',28);
%         set(gca, 'FontSize', 16, 'FontName', 'Times');
%         subplot(3,1,3); hold on;
%         b = boxplot(B_matrix_1se.',g,'Colors',cmap,'Symbol','');
%         set(b,{'linew'},{1})
%         plot([0,length(classkey)+1],[0,0],'k--','LineWidth',2);
%         set(gca,'XTick',1:length(subclasses));
%         set(gca,'XTickLabel',subclasses,'XTickLabelRotation',90);
%         ylim_ = max(abs([max(B_matrix_1se(:)),min(B_matrix_1se(:))]));
%         set(gca,'YTick',[-ylim_,ylim_])
%         text(-0.45,0,'0','FontSize',16,'FontName','Times')
%         ytickformat('%0.2f')
%         ylim([-1.1*ylim_,1.1*ylim_]);
%         ylabel('Regression Coefficients');
%         xlabel('');
%         set(gca,'TickLength',[0 0])
%         title(['\lambda_1_s_e Ridge Coefficients, ' datsetlabels{i}],'FontSize',28);
%         set(gca, 'FontSize', 16, 'FontName', 'Times');
%         if savenclose
%             cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
%             print(sprintf('RidgeCoefBar_Yao_%s',datsetname),'-dtiffn','-r300'); close;
% %             print(sprintf('RidgeCoefBar_%s',datsetname),'-dtiffn','-r300'); close;
%             cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis_Project/Nexis/;
%         end
%     end
% end
% if savenclose
%     cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
%     print(sprintf('Bootstrap_Lcurves_Yao_%d',kfold),'-dtiffn','-r300'); close;
% %     print(sprintf('Bootstrap_Lcurves_%d',kfold),'-dtiffn','-r300'); close;
%     cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis_Project/Nexis/;
% end
% 
% %% 5. Linear models, top correlations
% numsigtypes = 5;
% % numsigtypes = length(classkey);
% Y_ends = cell(1,length(datsetnames));
% sigtypes_corr_topn = cell(1,length(datsetnames));
% mdls_corr_topn = cell(1,length(datsetnames));
% Y_preds_corr_topn = mdls_corr_topn;
% subclasses_corr_topn = mdls_corr_topn;
% 
% % sigtypes_corr_bic = cell(1,length(datsetnames));
% % mdls_corr_bic = cell(1,length(datsetnames));
% % Y_preds_corr_bic = mdls_corr_bic;
% % subclasses_corr_bic = mdls_corr_bic;
% 
% sigtypes_parcorr_topn = cell(1,length(datsetnames));
% mdls_parcorr_topn = cell(1,length(datsetnames));
% Y_preds_parcorr_topn = mdls_parcorr_topn;
% subclasses_parcorr_topn = mdls_parcorr_topn;
% 
% % sigtypes_parcorr_bic = cell(1,length(datsetnames));
% % mdls_parcorr_bic = cell(1,length(datsetnames));
% % Y_preds_parcorr_bic = mdls_parcorr_bic;
% % subclasses_parcorr_bic = mdls_parcorr_bic;
% 
% sigtypes_coeff_bic = cell(1,length(datsetnames));
% mdls_coeff_bic = cell(1,length(datsetnames));
% Y_preds_coeff_bic = mdls_coeff_bic;
% subclasses_coeff_bic = mdls_coeff_bic;
% 
% plotting = 0;
% savenclose = 0;
% for i = 1:length(datsetnames)
%     % Establish top N types by corr. and partial corr.
%     [~,corrsortinds] = sort(abs(corrmat_all(i,:)),'descend');
%     corrinds_topn = sort(corrsortinds(1:numsigtypes));
%     sigtypes_corr_topn{i} = corrinds_topn;
%     [~,parcorrsortinds] = sort(abs(parcorrmat_all(i,:)),'descend');
%     parcorrinds_topn = sort(parcorrsortinds(1:numsigtypes));
%     sigtypes_parcorr_topn{i} = parcorrinds_topn;
% 
%     % Threshold by coefficient value from bootstrapped OLS, use only
%     % significant values for BIC model selection
%     sigtypeinds_B0_i = sigtypeinds_0{i};
%     subclasses_B0_i = subclasses(sigtypeinds_B0_i);
%     B_mean_0 = mean(B_0s{i},2);
%     B_mean_0_i = B_mean_0(sigtypeinds_B0_i);  
%     [~,coeffsortinds_B0] = sort(abs(B_mean_0_i),'descend');
%     subclasses_B0_i_sort = subclasses_B0_i(coeffsortinds_B0);
%     
%     % Fit models for top N types
%     datsetname = datsetnames{i};
%     Y = mousedata_struct.(datsetname).data;
%     naninds = isnan(Y(:,1));
%     Y_end = Y(~naninds,end);
%     Y_ends{i} = Y_end;
% %     Xct = Tasic_ng606(~naninds,:);
%     Xct = Yao_nG1300(~naninds,:);
%     Xct_sums = repmat(max(Xct),size(Xct,1),1);
%     Xct_norm = Xct./Xct_sums;
%     Xct_norm_corr_topn = Xct_norm(:,corrinds_topn);
%     mdls_corr_topn{i} = fitlm(Xct_norm_corr_topn,Y_end);
%     Y_preds_corr_topn{i} = [ones(length(Y_end),1), Xct_norm_corr_topn]...
%         * mdls_corr_topn{i}.Coefficients.Estimate;
%     subclasses_corr_topn{i} = subclasses(corrinds_topn);
%     Xct_norm_parcorr_topn = Xct_norm(:,parcorrinds_topn);
%     mdls_parcorr_topn{i} = fitlm(Xct_norm_parcorr_topn,Y_end);
%     Y_preds_parcorr_topn{i} = [ones(length(Y_end),1), Xct_norm_parcorr_topn]...
%         * mdls_parcorr_topn{i}.Coefficients.Estimate;
%     subclasses_parcorr_topn{i} = subclasses(parcorrinds_topn);
%     
%     % Fit models based on BIC
%     bic_calc = @(x,y,z) -2*x + (y * log(z));
%     bics_coeff = zeros(1,length(sigtypeinds_B0_i));
% %     bics_parcorr = zeros(1,size(corrmat_all,2));
%     for j = 1:length(sigtypeinds_B0_i)
%         subclasses_ij = subclasses_B0_i_sort(1:j); 
%         Xct_norm_coeff_j = Xct_norm(:,ismember(subclasses,subclasses_ij));
%         mdls_coeff_j = fitlm(Xct_norm_coeff_j,Y_end);
%         bics_coeff(j) = bic_calc(mdls_coeff_j.LogLikelihood,(j+1),length(Y_end));
% %         parcorrinds_j = sort(parcorrsortinds(1:j));
% %         Xct_norm_parcorr_j = Xct_norm(:,parcorrinds_j);
% %         mdls_parcorr_j = fitlm(Xct_norm_parcorr_j,Y_end);
% %         bics_parcorr(j) = bic_calc(mdls_parcorr_j.LogLikelihood,(j+1),length(Y_end));
%     end
%     if plotting
%         figure; scatter(1:length(sigtypeinds_B0_i),bics_coeff); title(datsetname); xlabel('n'); ylabel('BIC');
%     end
%     [~, modelind_coeff] = min(bics_coeff);
%     subclasses_coeff_bic{i} = sort(subclasses_B0_i_sort(1:modelind_coeff));
%     sigtypes_coeff_bic{i} = find(ismember(subclasses,subclasses_coeff_bic{i}));
%     Xct_norm_coeff_bic = Xct_norm(:,sigtypes_coeff_bic{i});
%     mdls_coeff_bic{i} = fitlm(Xct_norm_coeff_bic,Y_end);
%     Y_preds_coeff_bic{i} = [ones(length(Y_end),1), Xct_norm_coeff_bic]...
%         * mdls_coeff_bic{i}.Coefficients.Estimate;
% 
% 
% %     [~, modelind_parcorr] = min(bics_parcorr);
% %     parcorrinds_bic = sort(parcorrsortinds(1:modelind_parcorr));
% %     sigtypes_parcorr_bic{i} = parcorrinds_bic;
% %     Xct_norm_parcorr_bic = Xct_norm(:,parcorrinds_bic);
% %     mdls_parcorr_bic{i} = fitlm(Xct_norm_parcorr_bic,Y_end);
% %     Y_preds_parcorr_bic{i} = [ones(length(Y_end),1), Xct_norm_parcorr_bic]...
% %         * mdls_parcorr_bic{i}.Coefficients.Estimate;
% %     subclasses_parcorr_bic{i} = subclasses(parcorrinds_bic);
% end
% 
% if plotting
%     % Corr, top N
%     f = figure('Name','CorrTopN','units','inch','position',[0 0 15 11]);
%     cmap = hsv(length(datsetnames));
%     for i = 1:length(datsetnames)
%         subplot(3,4,i); % fix later for not 8 datasets
%         scatter(Y_preds_corr_topn{i},Y_ends{i},30,cmap(i,:),shapes{i});
%         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
%         max_x = max(Y_preds_corr_topn{i}); max_y = max(Y_ends{i});
%         min_x = min(Y_preds_corr_topn{i}); min_y = min(Y_ends{i});
%         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
%         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
%         if round(min_x,2) ~= 0 
%             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
%                 num2str(max_x,'%.2f')});
%         else
%             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
%                 num2str(max_x,'%.2f')});
%         end
%         if strcmp(datsetnames{i},'Clavaguera')
%             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
%                     num2str(max_y,'%.2f')});
%         else
%             if round(min_y,2) ~= 0 
%                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
%                     num2str(max_y,'%.2f')});
%             else
%                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
%                     num2str(max_y,'%.2f')});
%             end
%         end
%         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
%         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls_corr_topn{i}.Rsquared.Adjusted),...
%             'FontName','Times','FontSize',16);
%         title(datsetlabels{i},'FontSize',20);
%         set(gca,'FontSize',16,'FontName','Times');
%     end
%     han=axes(f,'visible','off'); 
%     han.XLabel.Visible='on';
%     han.YLabel.Visible='on';
%     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
%         'FontName','Times','FontWeight','bold');
%     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
%     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
%         'FontName','Times','FontWeight','bold');
%     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
%     if savenclose
%         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
%         print(sprintf('LinearModelPlots_Yao_Corrs_Top%dTypes',numsigtypes),'-dtiffn','-r300'); close;
% %         print(sprintf('LinearModelPlots_Corrs_Top%dTypes',numsigtypes),'-dtiffn','-r300'); close;
%         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis/;
%     end
% %     % Partial corr, top N
% %     f = figure('Name','ParCorrTopN','units','inch','position',[0 0 15 7]);
% %     cmap = hsv(length(datsetnames));
% %     for i = 1:length(datsetnames)
% %         subplot(2,4,i); % fix later for not 8 datasets
% %         scatter(Y_preds_parcorr_topn{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
% %         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
% %         max_x = max(Y_preds_parcorr_topn{i}); max_y = max(Y_ends{i});
% %         min_x = min(Y_preds_parcorr_topn{i}); min_y = min(Y_ends{i});
% %         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
% %         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
% %         if round(min_x,2) ~= 0 
% %             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         else
% %             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         end
% %         if strcmp(datsetnames{i},'Clavaguera')
% %             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %         else
% %             if round(min_y,2) ~= 0 
% %                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             else
% %                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             end
% %         end
% %         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
% %         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls_parcorr_topn{i}.Rsquared.Adjusted),...
% %             'FontName','Times','FontSize',16);
% %         title(datsetlabels{i},'FontSize',20);
% %         set(gca,'FontSize',16,'FontName','Times');
% %     end
% %     han=axes(f,'visible','off'); 
% %     han.XLabel.Visible='on';
% %     han.YLabel.Visible='on';
% %     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
% %     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
% %     if savenclose
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
% %         print(sprintf('LinearModelPlots_ParCorrs_Top%dTypes',numsigtypes),'-dtiffn','-r300'); close;
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis/;
% %     end
% % 
%     % Corr, BIC
%     f = figure('Name','CorrBIC','units','inch','position',[0 0 15 7]);
%     cmap = hsv(length(datsetnames));
%     for i = 1:length(datsetnames)
%         subplot(3,4,i+1); % fix later for not 8 datasets
%         scatter(Y_preds_coeff_bic{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
%         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
%         max_x = max(Y_preds_coeff_bic{i}); max_y = max(Y_ends{i});
%         min_x = min(Y_preds_coeff_bic{i}); min_y = min(Y_ends{i});
%         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
%         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
%         if round(min_x,2) ~= 0 
%             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
%                 num2str(max_x,'%.2f')});
%         else
%             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
%                 num2str(max_x,'%.2f')});
%         end
%         if strcmp(datsetnames{i},'Clavaguera')
%             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
%                     num2str(max_y,'%.2f')});
%         else
%             if round(min_y,2) ~= 0 
%                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
%                     num2str(max_y,'%.2f')});
%             else
%                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
%                     num2str(max_y,'%.2f')});
%             end
%         end
%         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
%         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls_coeff_bic{i}.Rsquared.Adjusted),...
%             'FontName','Times','FontSize',16);
%         title(datsetlabels{i},'FontSize',20);
%         set(gca,'FontSize',16,'FontName','Times');
%     end
%     han=axes(f,'visible','off'); 
%     han.XLabel.Visible='on';
%     han.YLabel.Visible='on';
%     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
%         'FontName','Times','FontWeight','bold');
%     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
%     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
%         'FontName','Times','FontWeight','bold');
%     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
%     if savenclose
%         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
%         print('LinearModelPlots_Coeffs_BIC','-dtiffn','-r300'); close;
%         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis/;
%     end
% % 
% %     % Partial Corr, BIC
% %     f = figure('Name','ParCorrBIC','units','inch','position',[0 0 15 7]);
% %     cmap = hsv(length(datsetnames));
% %     for i = 1:length(datsetnames)
% %         subplot(2,4,i); % fix later for not 8 datasets
% %         scatter(Y_preds_parcorr_bic{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
% %         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
% %         max_x = max(Y_preds_parcorr_bic{i}); max_y = max(Y_ends{i});
% %         min_x = min(Y_preds_parcorr_bic{i}); min_y = min(Y_ends{i});
% %         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
% %         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
% %         if round(min_x,2) ~= 0 
% %             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         else
% %             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         end
% %         if strcmp(datsetnames{i},'Clavaguera')
% %             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %         else
% %             if round(min_y,2) ~= 0 
% %                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             else
% %                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             end
% %         end
% %         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
% %         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls_parcorr_bic{i}.Rsquared.Adjusted),...
% %             'FontName','Times','FontSize',16);
% %         title(datsetlabels{i},'FontSize',20);
% %         set(gca,'FontSize',16,'FontName','Times');
% %     end
% %     han=axes(f,'visible','off'); 
% %     han.XLabel.Visible='on';
% %     han.YLabel.Visible='on';
% %     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
% %     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
% %     if savenclose
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
% %         print('LinearModelPlots_ParCorrs_BIC','-dtiffn','-r300'); close;
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis/;
% %     end    
% end
% 
% %% 5. Brainframe images
% addpath('/Users/justintorok/Documents/MATLAB/Brainframe-Dev/Brainframe')
% types_bf = {'Lamp5','L2_3_IT','Oligo'};
% datasets_bf = {'IbaHippInj','Hurtado','DS9'};
% 
% nsig = cell(42,2);
% nsig(:,1) = subclasses.';
% for i = 1:length(subclasses)
%     nsig{i,2} = 0;
% end
% 
% for i = 1:length(subclasses_corr)
%     inds = find(ismember(subclasses,subclasses_corr{i}));
%     for j = 1:length(inds)
%         nsig{inds(j),2} = nsig{inds(j),2} + 1/length(subclasses_corr);
%     end
% end
% nsigvec = cell2mat(nsig(:,2));
% [~,sort_inds] = sort(nsigvec,'descend');
% nsig = nsig(sort_inds,:);
% 
% nsigcorr = cell(42,2);
% nsigcorr(:,1) = subclasses.';
% for i = 1:length(subclasses)
%     nsigcorr{i,2} = 0;
% end
% 
% for i = 1:length(subclasses_corr_topn)
%     inds = find(ismember(subclasses,subclasses_corr_topn{i}));
%     for j = 1:length(inds)
%         nsigcorr{inds(j),2} = nsigcorr{inds(j),2} + 1/length(subclasses_corr_topn);
%     end
% end
% nsigvec = cell2mat(nsigcorr(:,2));
% [~,sort_inds] = sort(nsigvec,'descend');
% nsigcorr = nsigcorr(sort_inds,:);
% 
% 
% 
% % 
% % %% 5. Linear models for significant cell types
% % sigtypeinds_parcorr = sigtypes_partialp001;
% % mdls_parcorr_topn = cell(1,length(datsetnames));
% % Y_preds_parcorr_topn = mdls_parcorr_topn;
% % subclasses_parcorr_topn = mdls_parcorr_topn;
% % sigtypeinds_bootstrap = sigtypeinds_min; % OLS - can change, but OLS is sparsest and justifiable
% % mdls_bootstrap = cell(1,length(datsetnames));
% % Y_preds_bootstrap = mdls_bootstrap;
% % subclasses_bootstrap = mdls_bootstrap;
% % Y_ends = cell(1,length(datsetnames));
% % 
% % plotting = 1;
% % savenclose = 0;
% % for i = 1:length(datsetnames)
% %     datsetname = datsetnames{i};
% %     Y = mousedata_struct.(datsetname).data;
% %     naninds = isnan(Y(:,1));
% %     Y_end = Y(~naninds,end);
% %     Xct = Tasic_ng606(~naninds,:);
% %     Xct_sums = repmat(max(Xct),size(Xct,1),1);
% %     Xct_norm = Xct./Xct_sums;
% %     Xct_norm_parcorr_topn = Xct_norm(:,sigtypeinds_parcorr{i});
% %     mdls_parcorr_topn{i} = fitlm(Xct_norm_parcorr_topn,Y_end);
% %     Y_preds_parcorr_topn{i} = [ones(length(Y_end),1), Xct_norm_parcorr_topn]...
% %         * mdls_parcorr_topn{i}.Coefficients.Estimate;
% %     subclasses_parcorr_topn{i} = subclasses(sigtypeinds_parcorr{i});
% %     Xct_norm_bootstrap = Xct_norm(:,sigtypeinds_bootstrap{i});
% %     mdls_bootstrap{i} = fitlm(Xct_norm_bootstrap,Y_end);
% %     Y_preds_bootstrap{i} = [ones(length(Y_end),1), Xct_norm_bootstrap]...
% %         * mdls_bootstrap{i}.Coefficients.Estimate;
% %     subclasses_bootstrap{i} = subclasses(sigtypeinds_bootstrap{i});
% %     Y_ends{i} = Y_end;
% % end
% % 
% % if plotting
% %     f = figure('units','inch','position',[0 0 15 7]);
% %     cmap = hsv(length(datsetnames));
% %     for i = 1:length(datsetnames)
% %         subplot(2,4,i); % fix later for not 8 datasets
% %         scatter(Y_preds_parcorr_topn{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
% %         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
% %         max_x = max(Y_preds_parcorr_topn{i}); max_y = max(Y_ends{i});
% %         min_x = min(Y_preds_parcorr_topn{i}); min_y = min(Y_ends{i});
% %         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
% %         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
% %         if round(min_x,2) ~= 0 
% %             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         else
% %             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         end
% %         if strcmp(datsetnames{i},'Clavaguera')
% %             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %         else
% %             if round(min_y,2) ~= 0 
% %                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             else
% %                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             end
% %         end
% %         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
% %         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls_parcorr_topn{i}.Rsquared.Adjusted),...
% %             'FontName','Times','FontSize',16);
% %         title(datsetlabels{i},'FontSize',20);
% %         set(gca,'FontSize',16,'FontName','Times');
% %     end
% %     han=axes(f,'visible','off'); 
% %     han.XLabel.Visible='on';
% %     han.YLabel.Visible='on';
% %     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
% %     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
% %     if savenclose
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
% %         print('LinearModelPlots_PartialCorrs','-dtiffn','-r300'); close;
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis/;
% %     end
% %     
% %     f = figure('units','inch','position',[0 0 15 7]);
% %     cmap = hsv(length(datsetnames));
% %     for i = 1:length(datsetnames)
% %         subplot(2,4,i); % fix later for not 8 datasets
% %         scatter(Y_preds_bootstrap{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
% %         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
% %         max_x = max(Y_preds_bootstrap{i}); max_y = max(Y_ends{i});
% %         min_x = min(Y_preds_bootstrap{i}); min_y = min(Y_ends{i});
% %         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
% %         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
% %         if round(min_x,2) ~= 0 
% %             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         else
% %             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
% %                 num2str(max_x,'%.2f')});
% %         end
% %         if strcmp(datsetnames{i},'Clavaguera')
% %             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %         else
% %             if round(min_y,2) ~= 0 
% %                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             else
% %                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
% %                     num2str(max_y,'%.2f')});
% %             end
% %         end
% %         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
% %         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls_bootstrap{i}.Rsquared.Adjusted),...
% %             'FontName','Times','FontSize',16);
% %         title(datsetlabels{i},'FontSize',20);
% %         set(gca,'FontSize',16,'FontName','Times');
% %     end
% %     han=axes(f,'visible','off'); 
% %     han.XLabel.Visible='on';
% %     han.YLabel.Visible='on';
% %     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
% %     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
% %         'FontName','Times','FontWeight','bold');
% %     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
% %     if savenclose
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Figures/;
% %         print('LinearModelPlots_BootstrapOLS','-dtiffn','-r300'); close;
% %         cd /Users/justintorok/Documents/MATLAB/Nexis_Project/Nexis/;
% %     end
% % end
% % 
% % % %% 
% % % Y = mousedata_struct.Hurtado.data;
% % % naninds = isnan(Y(:,1));
% % % Y_end = Y(~naninds,end);
% % % Xct = Tasic_ng606(~naninds,:);
% % % mdl = mvregress(Xct,Y_end);
% % % 
% % % ranksmat = zeros(length(datsetnames),length(subclasses));
% % % for i = 1:length(datsetnames)
% % %     sortinds = sortindsmat(i,:);
% % %     for j = 1:length(subclasses)
% % %         ranksmat(i,j) = find(sortinds == j);
% % %     end
% % % end
% % % ranksmat_mean = mean(ranksmat);
% % % [~,ctranks] = sort(ranksmat_mean);
% % % ranksmat_rankorder = ranksmat(:,ctranks);
% % % subclasses_rankorder = subclasses(ctranks);
% % % 
% % % plotting = 'false';
% % % if logical(plotting)
% % %     figure('units','inch','position',[0 0 5 9]); 
% % %     cmap = flipud(pink(25));
% % %     xs = [1 length(datsetnames)]; ys = [1 length(subclasses)];
% % %     imagesc(xs,ys,ranksmat_rankorder.'); 
% % %     h = colorbar; set(h,'YDir','reverse');
% % %     colormap(cmap);
% % %     set(gca,'ytick',1:length(subclasses));
% % %     figlabs = subclasses_rankorder;
% % %     set(gca,'YTickLabel',figlabs);
% % %     set(gca,'xtick',1:length(datsetnames));
% % %     set(gca,'TickLength',[0 0]);
% % %     set(gca,'XAxisLocation','top');
% % %     set(gca,'XTickLabelRotation',45);
% % %     datsetlabels = cellfun(@(x) strrep(x,'_',' '), datsetnames,'UniformOutput',0);
% % %     set(gca,'XTickLabel',datsetlabels);
% % % %     ylabel('Cell Types','FontSize',20);
% % %     set(gca,'FontSize',16,'FontName','Times');
% % % end
% % % print('LASSORankings','-dpng','-r300'); close;
% % % 
% % % %% Linear models for top three cell types for each study (LASSO)
% % % mdls = cell(1,length(datsetnames));
% % % Y_ends = mdls;
% % % Y_preds = mdls;
% % % subclasses_top3 = mdls;
% % % for i = 1:length(datsetnames)
% % %     datsetname = datsetnames{i};
% % %     B = Bs{i}; FitInfo = FitInfos{i};
% % %     ind_top3 = find(FitInfo.DF==3,1,'last');
% % %     lambda = FitInfo.Lambda(ind_top3);
% % %     ct_top3 = top3_types{i};
% % %     ctinds_top3 = zeros(1,length(ct_top3));
% % %     for j = 1:length(ctinds_top3)
% % %         ctinds_top3(j) = find(ismember(classkey,ct_top3{j}));
% % %     end
% % %     subclasses_top3{i} = subclasses(ctinds_top3);
% % %     Y = mousedata_struct.(datsetname).data;
% % %     naninds = isnan(Y(:,1));
% % %     Y_end = Y(~naninds,end);
% % %     Xct = Tasic_ng606(~naninds,:);
% % %     Xct_sums = repmat(max(Xct),size(Xct,1),1);
% % %     Xct_norm = Xct./Xct_sums;
% % %     Xct_norm_top3 = Xct_norm(:,ctinds_top3);
% % %     mdls{i} = fitlm(Xct_norm_top3,Y_end);
% % %     Y_preds{i} = [ones(length(Y_end),1), Xct_norm_top3] * mdls{i}.Coefficients.Estimate;
% % %     Y_ends{i} = Y_end;
% % % end
% % % plotting = 'false';
% % % if logical(plotting)
% % %     f = figure('units','inch','position',[0 0 15 7]);
% % %     cmap = hsv(length(datsetnames));
% % %     datsetlabels = cellfun(@(x) strrep(x,'_',' '), datsetnames,'UniformOutput',0);
% % %     for i = 1:length(datsetnames)
% % %         subplot(2,4,i); % fix later for not 8 datasets
% % %         scatter(Y_preds{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
% % %         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
% % %         max_x = max(Y_preds{i}); max_y = max(Y_ends{i});
% % %         min_x = min(Y_preds{i}); min_y = min(Y_ends{i});
% % %         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
% % %         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
% % %         % xtickformat('%.1d'); ytickformat('%.1d'); 
% % %         if round(min_x,2) ~= 0 
% % %             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
% % %                 num2str(max_x,'%.2f')});
% % %         else
% % %             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
% % %                 num2str(max_x,'%.2f')});
% % %         end
% % %         if strcmp(datsetnames{i},'Clavaguera')
% % %             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
% % %                     num2str(max_y,'%.2f')});
% % %         else
% % %             if round(min_y,2) ~= 0 
% % %                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
% % %                     num2str(max_y,'%.2f')});
% % %             else
% % %                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
% % %                     num2str(max_y,'%.2f')});
% % %             end
% % %         end
% % %         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
% % %         subs_top3 = subclasses_top3{i};
% % %         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls{i}.Rsquared.Adjusted),...
% % %             'FontName','Times','FontSize',16);
% % %         title(datsetlabels{i},'FontSize',20);
% % %         set(gca,'FontSize',16,'FontName','Times');
% % %     end
% % %     han=axes(f,'visible','off'); 
% % %     han.XLabel.Visible='on';
% % %     han.YLabel.Visible='on';
% % %     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
% % %         'FontName','Times','FontWeight','bold');
% % %     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
% % %     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
% % %         'FontName','Times','FontWeight','bold');
% % %     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
% % %     print('Top3Scatters_ByLasso','-dpng','-r300'); close;
% % % end
% % % 
% % % %% Stepwise regression w.r.t. t(end) tau pathology 
% % % 
% % % % mdls_stepwise = cell(1,length(datsetnames));
% % % % top_celltypes_stepwise = mdls_stepwise;
% % % % sortindsmat = zeros(length(datsetnames),length(classkey));
% % % % for i = 1:length(datsetnames)
% % % %     datsetname = datsetnames{i};
% % % %     Y = mousedata_struct.(datsetname).data;
% % % %     naninds = isnan(Y(:,1));
% % % %     Y_end = Y(~naninds,end);
% % % %     Xct = Tasic_ng606(~naninds,:);
% % % %     Xct_sums = repmat(max(Xct),size(Xct,1),1);
% % % %     Xct_norm = Xct./Xct_sums;
% % % %     mdls_stepwise{i} = stepwiselm(Xct_norm,Y_end,'constant','Upper','linear','VarNames',...
% % % %         [subclasses 'Y_end'],'PEnter',0.01/length(subclasses),...
% % % %         'PRemove',0.05/length(subclasses));
% % % %     nonzeroinds = zeros(1,size(Xct,2));
% % % % end
% % % 
% % % %% Linear models for top three cell types for each study (Correlations)
% % % mdls = cell(1,length(datsetnames));
% % % Y_ends = mdls;
% % % Y_preds = mdls;
% % % subclasses_top3_corrs = mdls;
% % % for i = 1:length(datsetnames)
% % %     datsetname = datsetnames{i};
% % %     corrs_i = corrmat_all(i,:);
% % %     [~, inds_i] = sort(corrs_i,'descend');
% % %     subclasses_top3_corrs{i} = subclasses(inds_i(1:3));
% % %     Y = mousedata_struct.(datsetname).data;
% % %     naninds = isnan(Y(:,1));
% % %     Y_end = Y(~naninds,end);
% % %     Xct = Tasic_ng606(~naninds,:);
% % %     Xct_sums = repmat(max(Xct),size(Xct,1),1);
% % %     Xct_norm = Xct./Xct_sums;
% % %     Xct_norm_top3 = Xct_norm(:,inds_i(1:3));
% % %     mdls{i} = fitlm(Xct_norm_top3,Y_end);
% % %     Y_preds{i} = [ones(length(Y_end),1), Xct_norm_top3] * mdls{i}.Coefficients.Estimate;
% % %     Y_ends{i} = Y_end;
% % % end
% % % 
% % % if logical(plotting)
% % %     f = figure('units','inch','position',[0 0 15 7]);
% % %     cmap = hsv(length(datsetnames));
% % %     datsetlabels = cellfun(@(x) strrep(x,'_',' '), datsetnames,'UniformOutput',0);
% % %     for i = 1:length(datsetnames)
% % %         subplot(2,4,i); % fix later for not 8 datasets
% % %         scatter(Y_preds{i},Y_ends{i},50,cmap(i,:),shapes{i},'filled');
% % %         h = lsline; set(h,'LineWidth',2,'Color','k','LineStyle','--');
% % %         max_x = max(Y_preds{i}); max_y = max(Y_ends{i});
% % %         min_x = min(Y_preds{i}); min_y = min(Y_ends{i});
% % %         xlim([min_x max_x]); ylim([min_y max_y*1.05]);
% % %         xticks([min_x (max_x+min_x)/2 max_x]); yticks([min_y (max_y+min_y)/2 max_y]);
% % %         % xtickformat('%.1d'); ytickformat('%.1d'); 
% % %         if round(min_x,2) ~= 0 
% % %             xticklabels({num2str(min_x,'%.2f'),num2str((max_x+min_x)/2,'%.2f'),...
% % %                 num2str(max_x,'%.2f')});
% % %         else
% % %             xticklabels({'0',num2str((max_x+min_x)/2,'%.2f'),...
% % %                 num2str(max_x,'%.2f')});
% % %         end
% % %         if strcmp(datsetnames{i},'Clavaguera')
% % %             yticklabels({num2str(min_y,'%.3f'),num2str((max_y+min_y)/2,'%.2f'),...
% % %                     num2str(max_y,'%.2f')});
% % %         else
% % %             if round(min_y,2) ~= 0 
% % %                 yticklabels({num2str(min_y,'%.2f'),num2str((max_y+min_y)/2,'%.2f'),...
% % %                     num2str(max_y,'%.2f')});
% % %             else
% % %                 yticklabels({'0',num2str((max_y+min_y)/2,'%.2f'),...
% % %                     num2str(max_y,'%.2f')});
% % %             end
% % %         end
% % %         % ax = gca; ax.YAxis.Exponent = -1; ax.XAxis.Exponent = -1;
% % %         subs_top3 = subclasses_top3{i};
% % %         text(0.05*(max_x-min_x)+min_x, 0.925*(max_y-min_y)+min_y, sprintf('R^2 = %.2f',mdls{i}.Rsquared.Adjusted),...
% % %             'FontName','Times','FontSize',16);
% % %         title(datsetlabels{i},'FontSize',20);
% % %         set(gca,'FontSize',16,'FontName','Times');
% % %     end
% % %     han=axes(f,'visible','off'); 
% % %     han.XLabel.Visible='on';
% % %     han.YLabel.Visible='on';
% % % %     ylh = ylabel(han,'Observed Pathology at End Time Point','FontSize',20,...
% % % %         'FontName','Times','FontWeight','bold');
% % % %     ylh.Position(1) = ylh.Position(1) - abs(ylh.Position(1) * 0.75);
% % % %     xlh = xlabel(han,'Predicted Pathology at End Time Point','FontSize',20,...
% % % %         'FontName','Times','FontWeight','bold');
% % % %     xlh.Position(2) = xlh.Position(2) - abs(xlh.Position(2) * 0.5);
% % % %     print('Top3Scatters_ByCorr','-dpng','-r300'); close;
% % % % end
% % % % 
% % % % %% Cell Type Cross-Correlation Matrix
% % %% 7c. Hurtado Brainframe, sphere
% brainframedir = '/Users/justintorok/Documents/MATLAB/Brainframe-Dev/Brainframe';
% addpath(brainframedir)
% datset = 'Hurtado';
% tpts = mousedata_struct.(datset).time_stamps;
% savenclose = 0;
% reggroups = zeros(213,1); %Chunk of code to define region_groups
% amy = 1:11; cer = 12:23; sub = 24:26; hip = 27:37; hyp = 38:57; 
% ncx = 58:95; med = 96:120; mid = 121:141; olf = 142:149; pal = 150:157;
% pon = 158:170; str = 171:178; tha = 179:213;
% reggroups(amy) = 1; reggroups(cer) = 2; reggroups(sub) = 3; 
% reggroups(hip) = 4; reggroups(hyp) = 5; reggroups(ncx) = 6;
% reggroups(med) = 7; reggroups(mid) = 8; reggroups(olf) = 9;
% reggroups(pal) = 10; reggroups(pon) = 11; reggroups(str) = 12;
% reggroups(tha) = 13;
% reggroups = [reggroups;reggroups];
% cmap = hsv(length(unique(reggroups))); %Creating colormap
% 
% for i = 1:length(tpts)
%     datinput = mousedata_struct.(datset).data(:,i);
%     nany = isnan(datinput);
%     datinput(nany) = 0;
%     xfac = 1.5*(max(datinput)/max(mousedata_struct.(datset).data(~nany,1)))^0.5;
%     imglab = sprintf('PathologyBrainframe_Data_%s_t%d_%s_sphere',datset,tpts(i),ctdataset);
%     input_struct_data = brainframe_inputs_mouse(brainframedir,'data',datinput,...
%                                                 'voxUreg',1,...
%                                                 'xfac',xfac,...
%                                                 'pointsize',10,...
%                                                 'norm_method','max',...
%                                                 'bgcolor','w',...
%                                                 'img_format','tiffn',...
%                                                 'cmap',cmap,...
%                                                 'region_groups',reggroups,...
%                                                 'sphere',1,...
%                                                 'sphere_npts',20,...
%                                                 'regsUbins',0,...
%                                                 'img_directory',figdirectory,...
%                                                 'img_labels',imglab,...
%                                                 'img_renderer','painters',...
%                                                 'savenclose',savenclose);
%     brainframe(input_struct_data);
% 
%     mdlpred = mdls_bic_hurtado{i}.Fitted;
%     mdlpred_1 = mdls_bic_hurtado{1}.Fitted;
%     mdlpred_norm = (mdlpred - min(mdlpred));
%     datinput = zeros(426,1);
%     datinput(~nany) = mdlpred_norm;
%     xfac = 1.5*(max(mdlpred)/max(mdlpred_1))^0.5;
%     imglab = sprintf('PathologyBrainframe_Predicted_%s_t%d_%s_sphere',datset,tpts(i),ctdataset);
%     input_struct_pred = brainframe_inputs_mouse(brainframedir,'data',datinput,...
%                                                 'voxUreg',1,...
%                                                 'xfac',xfac,...
%                                                 'pointsize',10,...
%                                                 'norm_method','max',...
%                                                 'bgcolor','w',...
%                                                 'img_format','tiffn',...
%                                                 'cmap',cmap,...
%                                                 'region_groups',reggroups,...
%                                                 'sphere',1,...
%                                                 'sphere_npts',20,...
%                                                 'regsUbins',0,...
%                                                 'img_directory',figdirectory,...
%                                                 'img_labels',imglab,...
%                                                 'img_renderer','painters',...
%                                                 'savenclose',savenclose);
%     brainframe(input_struct_pred);
% end

%% S1. Run Nexis:global on all tau models and filter out low-performing datasets
% load([matdir filesep 'eNDM_mousedata.mat'],'data426');
% studylist = fieldnames(data426).';
% studylist(ismember(studylist,{'BolundaCBD','BolundaDSAD'})) = []; 
% load([matdir filesep 'KaufmanDiamond_datasets_dat&seed.mat'],'data426');
% studylist = [studylist fieldnames(data426).'];
% R2s = zeros(1,length(studylist));
% mousedata_struct = struct;
% rng(0);
% 
% for i = 1:length(studylist)
%     outputs_ndm = stdNDM_mouse('study',studylist{i},'bootstrapping',0,...
%         'w_dir',0,'volcorrect',1);
%     R2s(i) = outputs_ndm.ndm.Full.results.lm_Rsquared_adj;
%     mousedata_struct.(studylist{i}) = outputs_ndm.ndm.Full;
% end
% thresh = 0.25; % R across all timepoints > 0.5
% datsetnames = studylist(R2s > thresh);
% for i = 1:length(studylist)
%     if ~ismember(studylist{i},datsetnames)
%         mousedata_struct = rmfield(mousedata_struct,studylist{i});
%     end
% end
% save([matdir filesep 'OutputStruct_CTCorrsReg_allDS.mat']);

%% S2. Genetic vulnerability signature, end timepoint, DE of top 2 cell types
% t = 'end';
% corrmat_t = CorrelationCalculator_singletimepoint(datsetnames,t,...
%     mousedata_struct,ctmat,corrtype);
% corrmat_end_mean = mean(corrmat_t);
% [~,maxind] = max(corrmat_end_mean);
% [~,minind] = min(corrmat_end_mean);
% toptypeinds = [maxind,minind];
% 
% thresh = 1;
% n_g = 100;
% types = subclasses(toptypeinds);
% de_genes_cts = DifferentialGeneExpressionIdentifier_CTs(genevct,gene_names,...
%     thresh,types,subclasses,n_g);
% plotting = 0;
% savenclose = 0;
% corrtype = 'Pearson';
% 
% for i = 1:length(toptypeinds)
%     genelist_inds = ismember(gene_names,de_genes_cts{2,i});
%     genemat_list = voxvgene(:,genelist_inds);
%     [~,genemat_list] = Voxel_To_Region(genemat_list,datapath);
%     
%     cmap = 'cool';
%     corrmat_t_gene = CorrelationCalculator_singletimepoint(datsetnames,t,...
%         mousedata_struct,genemat_list,corrtype);
%     if plotting
%         CorrelationBarPlot(corrmat_t_gene,datsetnames,de_genes_cts{2,i},corrtype,...
%             cmap,savenclose,figdirectory);
%     end
% end